spec/controllers/ops_controller/settings/upload_spec.rb

```diff
@@ -3,6 +3,166 @@
     stub_user(:features => :all)
   end
 
+  describe '#upload_png' do
+    let(:dir) { File.dirname(__FILE__) }
+
+    before do
+      allow(controller).to receive(:load_edit).and_return(true)
+      allow(controller).to receive(:redirect_to)
+      controller.instance_variable_set(:@sb, :active_tab => 'settings_custom_logos', :selected_server_id => 123)
+    end
+
+    it 'adds success flash message to @flash_array regarding updating custom logo image with a .png file' do
+      file = dir << '/data/test.png'
+      temp = Tempfile.new(['RackMultipart', '.png'])
+      png = File.open(file)
+      encoded_png = Base64.encode64(File.read(png.path))
+      File.open(temp.path, "wb") do |f|
+        f.write(Base64.decode64(encoded_png))
+      end
+      controller.params = {:upload => {:logo => ActionDispatch::Http::UploadedFile.new(:tempfile => temp, :filename => 'test.png', :type => 'image/png')}}
```

Comment:
I'm not sure I understand why it's read encoded then written decoded. That seems effectively the same as just copying the file.  Additionally, `ActionDispatch::Http::UploadedFile`, can work with any file IO (it doesn't have to be a literal temp file.  So I think this can be simplified to just

```suggestion
      file = File.join(__dir__, 'data/test.png')
      controller.params = {:upload => {:logo => ActionDispatch::Http::UploadedFile.new(:tempfile => File.open(file), :filename => 'test.png', :type => 'image/png')}}
```

Reply:
Unless you wanted to make sure the code work with an actual tempfile name with the "'RackMultipart', '.png'" parts?

Reply:
For this part I was just trying to replicate how the controller was receiving the data when using the UI. I was having trouble getting it to run so I thought that was the best way to get it working. Your code looks much simpler so I can give that a try. Also, the encoding and decoding was just because I thought that was the only way to copy a png file content to a temp file. Since your code doesn't need the temp file we wouldn't need that anymore then.